Bug: "RuntimeWarning: invalid value encountered in dot"

**Describe the bug**
I was letting the VE run for the full 24 steps untruncated over the weekend to test some code I was developing, but it appears to have run into an error at some point where the output terminal was absolutely spammed with the message `pyrealm\env_factors.py:199: RuntimeWarning: invalid value encountered in dot` as shown in the terminal snippet below. I eventually used ctrl+c to end the process because I had no idea what was happening and spamming output like that makes me think the code is glitching out somewhere. This should be the code running out the box on Python 3.14, so it seems strange to me that this is occurring.

**Additional comment from @jacobcook1995 :**

> _The underlying issue is that if the scipy integrator that the soil model uses is provided with invalid inputs (either from the input values or from the other models) the integrator doesn't crash but instead just continues to try and solve an unsolvable ODE set forever._
> _I already set up functionality to catch invalid inputs, but it looks like there some gaps in this, which I should address. (This wouldn't fix the underlying issue which I'm like 80% sure originates in another models code, but it would mean it would fail gracefully rather than iterating forever)_

**Versions:**
- Virtual Ecosystem: 0.1.1a16
- Python: 3.14
- OS: Windows 11

**Code I used to run VE:**
```python
ve_run_args = [
        "data/ve_example/config",
        "--out",
        "data/ve_example/out",
        "--logfile", 
        "data/ve_example/logfile.log"
    ]
ve_run_cli(ve_run_args)
```

**Extended terminal output:**
```shell
pyrealm\env_factors.py:199: RuntimeWarning: invalid value encountered in dot
pyrealm\env_factors.py:199: RuntimeWarning: invalid value encountered in dot
pyrealm\env_factors.py:199: RuntimeWarning: invalid value encountered in dot
pyrealm\env_factors.py:199: RuntimeWarning: invalid value encountered in dot
pyrealm\env_factors.py:199: RuntimeWarning: invalid value encountered in dot
pyrealm\env_factors.py:199: RuntimeWarning: invalid value encountered in dot
pyrealm\env_factors.py:199: RuntimeWarning: invalid value encountered in dot
pyrealm\env_factors.py:199: RuntimeWarning: invalid value encountered in dot
pyrealm\env_factors.py:199: RuntimeWarning: invalid value encountered in dot
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "C:\Users\okirk\AppData\Local\Python\pythoncore-3.14-64\Lib\cProfile.py", line 205, in <module>
    main()
    ~~~~^^
  File "C:\Users\okirk\AppData\Local\Python\pythoncore-3.14-64\Lib\cProfile.py", line 194, in main
    runctx(code, globs, None, options.outfile, options.sort)
    ~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\okirk\AppData\Local\Python\pythoncore-3.14-64\Lib\cProfile.py", line 20, in runctx
    return _pyprofile._Utils(Profile).runctx(statement, globals, locals,
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                                             filename, sort)
                                             ^^^^^^^^^^^^^^^
  File "C:\Users\okirk\AppData\Local\Python\pythoncore-3.14-64\Lib\profile.py", line 63, in runctx
    prof.runctx(statement, globals, locals)
    ~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\okirk\AppData\Local\Python\pythoncore-3.14-64\Lib\cProfile.py", line 103, in runctx
    exec(cmd, globals, locals)
    ~~~~^^^^^^^^^^^^^^^^^^^^^^
  File "run.py", line 29, in <module>
    ve_run_cli(ve_run_args)
    ~~~~~~~~~~^^^^^^^^^^^^^
  File "C:\Users\okirk\Documents\CODE\ve_project\ve3_14\ve3_14\Lib\site-packages\virtual_ecosystem\entry_points.py", line 261, in ve_run_cli
    ve_run(
    ~~~~~~^
        cfg_paths=args.cfg_paths,
        ^^^^^^^^^^^^^^^^^^^^^^^^^
    ...<2 lines>...
        progress=progress,
        ^^^^^^^^^^^^^^^^^^
    )
    ^
  File "C:\Users\okirk\Documents\CODE\ve_project\ve3_14\ve3_14\Lib\site-packages\virtual_ecosystem\main.py", line 322, in ve_run
    model.update(time_index)
    ~~~~~~~~~~~~^^^^^^^^^^^^
  File "C:\Users\okirk\Documents\CODE\ve_project\ve3_14\ve3_14\Lib\site-packages\virtual_ecosystem\core\base_model.py", line 438, in update
    self._update(time_index, **kwargs)
    ~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\okirk\Documents\CODE\ve_project\ve3_14\ve3_14\Lib\site-packages\virtual_ecosystem\models\soil\soil_model.py", line 358, in _update
    updated_soil_pools = self.integrate()
  File "C:\Users\okirk\Documents\CODE\ve_project\ve3_14\ve3_14\Lib\site-packages\virtual_ecosystem\models\soil\soil_model.py", line 494, in integrate
    output = solve_ivp(
        construct_full_soil_model,
    ...<14 lines>...
        ),
    )
  File "C:\Users\okirk\Documents\CODE\ve_project\ve3_14\ve3_14\Lib\site-packages\scipy\integrate\_ivp\ivp.py", line 657, in solve_ivp
    message = solver.step()
  File "C:\Users\okirk\Documents\CODE\ve_project\ve3_14\ve3_14\Lib\site-packages\scipy\integrate\_ivp\base.py", line 201, in step
    success, message = self._step_impl()
                       ~~~~~~~~~~~~~~~^^
  File "C:\Users\okirk\Documents\CODE\ve_project\ve3_14\ve3_14\Lib\site-packages\scipy\integrate\_ivp\rk.py", line 144, in _step_impl
    y_new, f_new = rk_step(self.fun, t, y, self.f, h, self.A,
                   ~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                           self.B, self.C, self.K)
                           ^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\okirk\Documents\CODE\ve_project\ve3_14\ve3_14\Lib\site-packages\scipy\integrate\_ivp\rk.py", line 64, in rk_step
    K[s] = fun(t + c * h, y + dy)
           ~~~^^^^^^^^^^^^^^^^^^^
  File "C:\Users\okirk\Documents\CODE\ve_project\ve3_14\ve3_14\Lib\site-packages\scipy\integrate\_ivp\base.py", line 158, in fun
    return self.fun_single(t, y)
           ~~~~~~~~~~~~~~~^^^^^^
  File "C:\Users\okirk\Documents\CODE\ve_project\ve3_14\ve3_14\Lib\site-packages\scipy\integrate\_ivp\base.py", line 24, in fun_wrapped
    return np.asarray(fun(t, y), dtype=dtype)
                      ~~~^^^^^^
  File "C:\Users\okirk\Documents\CODE\ve_project\ve3_14\ve3_14\Lib\site-packages\scipy\integrate\_ivp\ivp.py", line 595, in fun
    return fun(t, x, *args)
  File "C:\Users\okirk\Documents\CODE\ve_project\ve3_14\ve3_14\Lib\site-packages\virtual_ecosystem\models\soil\soil_model.py", line 916, in construct_full_soil_model
    return soil_pools.calculate_all_pool_updates(
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
        delta_pools_ordered=delta_pools_ordered,
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    ...<3 lines>...
        top_soil_layer_thickness=top_soil_layer_thickness,
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "C:\Users\okirk\Documents\CODE\ve_project\ve3_14\ve3_14\Lib\site-packages\virtual_ecosystem\models\soil\pools.py", line 524, in calculate_all_pool_updates
    enzyme_mediated = calculate_enzyme_mediated_rates(
        pools=self.pools,
    ...<2 lines>...
        enzyme_classes=self.enzyme_classes,
    )
  File "C:\Users\okirk\Documents\CODE\ve_project\ve3_14\ve3_14\Lib\site-packages\virtual_ecosystem\models\soil\pools.py", line 1134, in calculate_enzyme_mediated_rates
    calculate_enzyme_mediated_decomposition(
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
        soil_c_pool=getattr(pools, f"soil_cnp_pool_{substrate}_carbon"),
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    ...<3 lines>...
        enzyme_class=enzyme_classes[f"{source}_{substrate}"],
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
KeyboardInterrupt
 17%|██████████████▋                                                                         | 4/24 [19:53<1:39:28, 298.40s/it]
^CTraceback (most recent call last):
  File "C:\Users\okirk\Documents\CODE\ve_project\ve3_14\main.py", line 52, in <module>
    subprocess.run(command.split(), shell=True)
    ~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\okirk\AppData\Local\Python\pythoncore-3.14-64\Lib\subprocess.py", line 556, in run
    stdout, stderr = process.communicate(input, timeout=timeout)
                     ~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\okirk\AppData\Local\Python\pythoncore-3.14-64\Lib\subprocess.py", line 1212, in communicate
    self.wait()
    ~~~~~~~~~^^
  File "C:\Users\okirk\AppData\Local\Python\pythoncore-3.14-64\Lib\subprocess.py", line 1278, in wait
    return self._wait(timeout=timeout)
           ~~~~~~~~~~^^^^^^^^^^^^^^^^^
  File "C:\Users\okirk\AppData\Local\Python\pythoncore-3.14-64\Lib\subprocess.py", line 1604, in _wait
    result = _winapi.WaitForSingleObject(self._handle,
                                         timeout_millis)
KeyboardInterrupt```

Okay, I think I've found the source of the problem in the soil model. At present I check for NaN values in inputs to the soil integration, but I don't check for infinite values. It turns out that the other models can generate infinite values (not just NaN values) when things go wrong, and feeding these values into the soil integration causes it to evaluate forever.

When I extend the checking to also catch infinite values, the simulation now ends with the following:

```
[INFO] - base_model - update(437) - Updating soil model
[ERROR] - soil_model - integrate(490) - Soil model integration cannot proceed because the following variables have unexpected NaN values: {'soil_temperature'}
```

(Updating the message is still a TODO).

Fixing this basically just shifts your problem, but I guess it makes most sense to close this issue once I've made the fix, and then make a new issue for the downstream problems.

Once I've merged the fix in it's probably worth you `pip` installing the current state of develop, via `pip install --upgrade git+https://github.com/ImperialCollegeLondon/virtual_ecosystem.git@develop`, so that you have access to the most recent changes (i.e. including the fix)

--- a/tests/models/soil/test_soil_model.py
+++ b/tests/models/soil/test_soil_model.py
@@ -574,7 +574,7 @@ def test_integrate_with_nans(caplog, fixture_soil_model):
         (
             ERROR,
             "Soil model integration cannot proceed because the following variables "
-            "have unexpected NaN values: {'pH'}",
+            "contain invalid values (e.g. NaN or Inf): {'pH'}",
         ),
     )
 
@@ -662,7 +662,7 @@ def test_order_independance(
 
 
 @pytest.mark.parametrize(
-    argnames=["unexpected_nans", "variable_name", "input_data"],
+    argnames=["invalid_values", "variable_name", "input_data"],
     argvalues=[
         pytest.param(
             False,
@@ -676,22 +676,34 @@ def test_order_independance(
             DataArray([3.3, np.nan, 5.6, 7.9], dims=["cell_id"]),
             id="NaN",
         ),
+        pytest.param(
+            True,
+            "pH",
+            DataArray([3.3, np.inf, 5.6, 7.9], dims=["cell_id"]),
+            id="Inf",
+        ),
+        pytest.param(
+            True,
+            "pH",
+            DataArray([3.3, -np.inf, 5.6, 7.9], dims=["cell_id"]),
+            id="negative inf",
+        ),
     ],
 )
-def test_check_for_unexpected_nan_value_flat(
-    fixture_soil_model, unexpected_nans, variable_name, input_data
+def test_check_for_invalid_input_values_flat(
+    fixture_soil_model, invalid_values, variable_name, input_data
 ):
     """Test unexpected NaN checking values works for variables without layers."""
 
     fixture_soil_model.data[variable_name] = input_data
 
-    assert unexpected_nans == fixture_soil_model.check_for_unexpected_nan_values(
+    assert invalid_values == fixture_soil_model.check_for_invalid_input_values(
         var=variable_name
     )
 
 
 @pytest.mark.parametrize(
-    argnames=["unexpected_nans", "variable_name", "layer_name", "input_data"],
+    argnames=["invalid_values", "variable_name", "layer_name", "input_data"],
     argvalues=[
         pytest.param(
             False,
@@ -708,6 +720,13 @@ def test_check_for_unexpected_nan_value_flat(
             id="surface, bad",
         ),
         pytest.param(
+            True,
+            "air_temperature",
+            "index_surface",
+            np.array([3.3, np.inf, 5.6, 7.9]),
+            id="surface, inf",
+        ),
+        pytest.param(
             False,
             "soil_temperature",
             "index_all_soil",
@@ -719,14 +738,21 @@ def test_check_for_unexpected_nan_value_flat(
             "soil_temperature",
             "index_all_soil",
             np.array([[3.3, 4.3, 5.6, 7.9], [np.nan, 26.1, 24.4, 29.8]]),
-            id="soil, bad",
+            id="soil, nan",
+        ),
+        pytest.param(
+            True,
+            "soil_temperature",
+            "index_all_soil",
+            np.array([[3.3, 4.3, 5.6, 7.9], [np.inf, 26.1, 24.4, 29.8]]),
+            id="soil, inf",
         ),
     ],
 )
-def test_check_for_unexpected_nan_value_layered(
+def test_check_for_invalid_input_values_layered(
     fixture_soil_model,
     fixture_core_components,
-    unexpected_nans,
+    invalid_values,
     variable_name,
     layer_name,
     input_data,
@@ -737,7 +763,7 @@ def test_check_for_unexpected_nan_value_layered(
     fixture_soil_model.data[variable_name] = lyr_str.from_template()
     fixture_soil_model.data[variable_name][getattr(lyr_str, layer_name)] = input_data
 
-    assert unexpected_nans == fixture_soil_model.check_for_unexpected_nan_values(
+    assert invalid_values == fixture_soil_model.check_for_invalid_input_values(
         var=variable_name
     )
 

--- a/virtual_ecosystem/models/soil/soil_model.py
+++ b/virtual_ecosystem/models/soil/soil_model.py
@@ -578,7 +578,7 @@ class SoilModel(
         else:
             subset = self.data[var]
 
-        return bool( ~ np.all(np.isfinite(subset)))
+        return bool(~np.all(np.isfinite(subset)))
 
     def convert_fruiting_body_production_to_rate(
         self, total_production: DataArray

--- a/virtual_ecosystem/models/soil/soil_model.py
+++ b/virtual_ecosystem/models/soil/soil_model.py
@@ -474,18 +474,18 @@ class SoilModel(
             **{name: np.array([]) for name in self.refreshed_variables},
         }
 
-        # Check if any values used by the soil model integration have NaN values (these
-        # can stall the integration)
-        unexpected_nans = set()
+        # Check if any values used by the soil model integration have invalid values
+        # (e.g. NaN or Inf values, as these can stall the integration)
+        invalid_values = set()
 
         for var in self.vars_required_for_update:
-            if self.check_for_unexpected_nan_values(var=var):
-                unexpected_nans.add(var)
+            if self.check_for_invalid_input_values(var=var):
+                invalid_values.add(var)
 
-        if unexpected_nans:
+        if invalid_values:
             to_raise_nan = ValueError(
                 "Soil model integration cannot proceed because the following "
-                f"variables have unexpected NaN values: {unexpected_nans}"
+                f"variables contain invalid values (e.g. NaN or Inf): {invalid_values}"
             )
             LOGGER.error(to_raise_nan)
             raise to_raise_nan
@@ -555,18 +555,20 @@ class SoilModel(
 
         return new_soil_pools | triplet_pools
 
-    def check_for_unexpected_nan_values(self, var: str) -> bool:
-        """Check if there are unexpected NaN values in the data for a specific variable.
+    def check_for_invalid_input_values(self, var: str) -> bool:
+        """Check if there are any invalid values in the data for a specific variable.
 
-        The soil model needs the air_temperature variable to have non-NaN values at the
-        soil surface, and the other layer structured variables to be defined for every
-        soil layer. For these variables, this function takes the appropriate subset.
+        This function checks for NaN and infinite values.
+
+        For variables with a layer structure, only the layers relevant to the soil model
+        are checked. For the air_temperature variable this is just the soil surface
+        layer, but for the other layer-structured variable all soil layers are checked.
 
         Args:
             var: The name of the variable being checked
 
         Returns:
-            Whether the data for the variable has any unexpected NaN values.
+            Whether the data for the variable has any invalid values.
         """
 
         if var == "air_temperature":
@@ -576,7 +578,7 @@ class SoilModel(
         else:
             subset = self.data[var]
 
-        return bool(subset.isnull().any())
+        return bool((subset.isnull() | np.isinf(subset)).any())
 
     def convert_fruiting_body_production_to_rate(
         self, total_production: DataArray

--- a/virtual_ecosystem/models/soil/soil_model.py
+++ b/virtual_ecosystem/models/soil/soil_model.py
@@ -578,7 +578,7 @@ class SoilModel(
         else:
             subset = self.data[var]
 
-        return bool((subset.isnull() | np.isinf(subset)).any())
+        return bool( ~ np.all(np.isfinite(subset)))
 
     def convert_fruiting_body_production_to_rate(
         self, total_production: DataArray

